pkg/services/alerting/eval_context.go

```diff
@@ -178,3 +181,51 @@ func getNewStateInternal(c *EvalContext) models.AlertStateType {
 
 	return models.AlertStateOK
 }
+
+// evaluateNotificationTemplateFields will treat the alert evaluation rule's name and message fields as Go text
+// templates, and execute the templates using data from the alert evaluation's tags
+func (c *EvalContext) evaluateNotificationTemplateFields() error {
+	if len(c.EvalMatches) < 1 {
+		return nil
+	}
+
+	msgTmpl, err := template.New("").Parse(c.Rule.Message)
+	if err != nil {
+		return err
+	}
+
+	templateDataMap := buildTemplateDataMap(c.EvalMatches)
+
+	var msgBuf bytes.Buffer
+	if err := msgTmpl.Execute(&msgBuf, templateDataMap); err != nil {
+		return err
+	}
+	c.Rule.Message = msgBuf.String()
+
+	titleTmpl, err := template.New("").Parse(c.Rule.Name)
+	if err != nil {
+		return err
+	}
+
+	var titleBuf bytes.Buffer
+	if err := titleTmpl.Execute(&titleBuf, templateDataMap); err != nil {
+		return err
+	}
+	c.Rule.Name = titleBuf.String()
+	return nil
+}
+
+// buildTemplateDataMap builds a map of alert evaluation tag names to a set of associated values (comma separated)
+func buildTemplateDataMap(evalMatches []*EvalMatch) map[string]string {
+	var result = map[string]string{}
+	for _, match := range evalMatches {
+		for tagName, tagValue := range match.Tags {
+			if _, exists := result[tagName]; exists && !strings.Contains(result[tagName], tagValue) {
```

Comment:
Isn't there a danger that `result[tagName]` contains a comma separated entry that is a super-string of `tagValue`?That is to say, I think you should instead use a precise regex here to make sure you don't get a partial match.

Reply:
Not 100% sure what you mean - maybe take a look at the tests I just added for this function and if you think there's a missing test case that highlights an issue, let me know!

Reply:
OK thx, will check the tests.